library/src/scala/compiletime/ops/string.scala

```diff
@@ -9,3 +9,32 @@ object string:
    *  @syntax markdown
    */
   type +[X <: String, Y <: String] <: String
+
+  /** Length of a `String` singleton type.
+   *  ```scala
+   *  val helloSize: Size["hello"] = 5
+   *  ```
+   *  @syntax markdown
+   */
+  type Length[X <: String] <: Int
+
+  /** Substring of a `String` singleton type, with a singleton type
+   * begin inclusive index `IBeg`, and a singleton type exclusive end index `IEnd`.
+   * The substring begins at the specified IBeg and extends to the character at index IEnd - 1.
+   * Thus the length of the substring is IEnd-IBeg.
+   *  ```scala
+   *  val x: Substring["hamburger", 4, 8] = "urge"
+   *  val y: Substring["smiles", 1, 5] = "mile"
+   *  ```
+   *  @syntax markdown
+   */
+  type Substring[S <: String, IBeg <: Int, IEnd <: Int] <: String
```

Comment:
Should we also add the `Substring` without an `IEnd`? Would overloading the type alias work?

Reply:
No, overloading will not work, so using the one with `IEnd` is the best we have. Those who need the short version can create an alias with `IEnd = Length[S]-1`.